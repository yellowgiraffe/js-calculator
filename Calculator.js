export default class Calculator {
  constructor() {
    this.currentInput = document.querySelector('#currentInput');
    this.previousInput = document.querySelector('#previousInput');
    this.signInput = document.querySelector('#sign');
    this.keys = document.querySelector('#buttons');
  }

  init() {
    this.reset();
    this.listenToButtons();
  }

  listenToButtons() {
    this.keys.addEventListener('click', (event) => {
      const { target } = event;

      if (!target.matches('button')) {
        return;
      }

      const buttonType = target.getAttribute('data-button');

      switch (buttonType) {
        case 'number':
          this.chooseNumber(target.textContent);
          break;
        case 'operator':
          this.chooseOperator(target.textContent);
          break;
        case 'equals':
          this.showResult();
          this.previousNumber = '';
          break;
        case 'clear':
          this.reset();
          break;
        default:
          return;
      }

      this.updateDisplay();
    });
  }

  updateDisplay() {
    this.currentInput.textContent = this.currentNumber;
    this.previousInput.textContent = this.previousNumber;
    this.signInput.textContent = this.operator;
  }

  chooseNumber(number) {
    if (this.currentNumber === '0') {
      this.currentNumber = '';
    }

    if (number === '.' && this.currentNumber.includes('.')) {
      return;
    }

    if (number === '.' && this.currentNumber === '') {
      this.currentNumber = '0.';
      return;
    }

    if (this.currentNumber === this.previousNumber && this.operator !== '') {
      this.currentNumber = '';
      this.currentNumber += number.toString();
    } else {
      this.currentNumber += number.toString();
    }
  }

  reset() {
    this.currentNumber = '0';
    this.previousNumber = '';
    this.operator = '';
  }

  chooseOperator(operation) {
    if (this.operator !== '') {
      this.showResult();
    }

    this.previousNumber = this.currentNumber;
    this.operator = operation;
    this.isWaitingForSecondOperand = true;
  }

  doMath() {
    let result;
    const prev = Number(this.previousNumber);
    const curr = Number(this.currentNumber);
    const sign = this.operator;

    switch (sign) {
      case '+':
        result = prev + curr;
        break;
      case '-':
        result = prev - curr;
        break;
      case 'ร':
        result = prev * curr;
        break;
      case 'รท':
        result = prev / curr;
        break;
      default:
        return;
    }

    if (Number.isNaN(result)) {
      return;
    }

    if (result.toString().length > 10) {
      return Number(result.toFixed(8));
    }

    return result;
  }

  showResult() {
    const result = this.doMath();
    this.currentNumber = result || '0';
    this.previousNumber = this.currentNumber;
    this.operator = '';
  }
}
