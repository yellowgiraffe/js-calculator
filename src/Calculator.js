/* eslint-disable consistent-return */

// TO DO
// Wyświetl komunikat "Nie można dzileić przez zero"
// Dodaj możliwość wpisywania ujemnej liczby jako drugi operand
// Zablokój klikanie przycisków + - * /, kiedy jeden był właśnie kliknięty
export default class Calculator {
  constructor() {
    this.currentInput = document.querySelector('#currentInput');
    this.previousInput = document.querySelector('#previousInput');
    this.signInput = document.querySelector('#sign');
    this.keys = document.querySelector('#buttons');
  }

  init() {
    this.reset();
    this.listenToButtons();
  }

  listenToButtons() {
    this.keys.addEventListener('click', event => {
      const { target } = event;

      if (!target.matches('button')) {
        return;
      }

      const buttonType = target.getAttribute('data-button');

      switch(buttonType) {
        case 'number':
          this.showInputNumbers(target.textContent);
          break;
        case 'operator':
          this.chooseOperator(target.textContent);
          break;
        case 'equals':
          this.equals();
          break;
        case 'clear':
          this.reset()
          break;
      }
    });
  }

  showInputNumbers(number) {
    if (this.currentInput.textContent === '0') {
      this.currentInput.textContent = '';
    }

    if (number === '.' && this.currentInput.textContent.includes('.')) {
      return;
    }

    if (number === '.' && this.currentInput.textContent === '') {
      this.currentInput.textContent = '0.';
      return;
    }

    if (number === '.' && this.currentInput.textContent === '-') {
      this.currentInput.textContent = '-0.';
      return;
    }

    // this.currentInput.textContent += number.toString();

    // To dziala dla jednej liczby
    // if (this.signInput.textContent === '') {
    //   this.currentInput.textContent += number.toString();
    // } 
    // else {
    //   this.currentInput.textContent = '';
    //   this.currentInput.textContent += number.toString();
    // }

    // Dziala dla dwóch cyfr, ale nie dziala 2 + 22
    if (this.currentInput.textContent === this.previousInput.textContent 
      && this.signInput.textContent !== '') {
        this.currentInput.textContent = '';
        this.currentInput.textContent += number.toString();
      } else {
        this.currentInput.textContent += number.toString();
      }
  }

  reset() {
    this.currentInput.textContent = '0';
    this.previousInput.textContent = '';
    this.signInput.textContent = '';
  }

  chooseOperator(operation) {
    if (this.currentInput.textContent === '0' && operation === '-') {
      this.currentInput.textContent = '-';
      return;
    }

    if (this.signInput.textContent !== '') {
      this.showResult();
    }

    this.previousInput.textContent = this.currentInput.textContent;
    this.signInput.textContent = operation;
  }

  doMath() {
    let result;
    const prev = Number(this.previousInput.textContent);
    const current = Number(this.currentInput.textContent);
    const operator = this.signInput.textContent;

    switch (operator) {
      case '+':
        result = prev + current;
        break;
      case '-':
        result = prev - current;
        break;
      case '×':
        result = prev * current;
        break;
      case '÷':
        result = prev / current;
        break;
      default:
        return;
    }

    if (Number.isNaN(result)) {
      return;
    }

    if (result.toString().length > 10) {
      return Number(result.toFixed(8));
    }

    return result;
  }

  showResult() {
    const result = this.doMath();
    this.previousInput.textContent = this.currentInput.textContent;
    this.currentInput.textContent = result || '0';
    this.signInput.textContent = '';
  }

  equals() {
    const result = this.doMath();
    console.log(result);
    this.previousInput.textContent += ` ${this.signInput.textContent} ${this.currentInput.textContent} =` ; // 9 + 1 =
    this.currentInput.textContent = result || '0';
    this.signInput.textContent = '';
  }
}
