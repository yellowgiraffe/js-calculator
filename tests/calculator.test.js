/* eslint-disable no-undef */
import fs from 'fs';
import { dirname } from 'path';
import { jest } from '@jest/globals';
import { fileURLToPath } from 'url';

import Calculator from '../src/Calculator';

const __dirname = dirname(fileURLToPath(import.meta.url));

const simulateClick = (el) => {
  el.dispatchEvent(new MouseEvent('click', { bubbles: true }));
}

let calc = null;
let html = null;

beforeAll(() => {
  html = fs.readFileSync(__dirname + '/../src/index.html', 'utf8');
});

beforeEach(() => {
  document.documentElement.innerHTML = html;
  calc = new Calculator();
});

describe('Class Calculator', () => {
  test('should exist and to be type of function', () => {
    expect(Calculator).toBeDefined();
    expect(typeof Calculator).toBe('function');
  });

  test('constructor should get DOM elements for display', () => {
    expect(calc.currentInput).toMatchSnapshot();
    expect(calc.previousInput).toMatchSnapshot();
    expect(calc.signInput).toMatchSnapshot();
    expect(calc.keys).toMatchSnapshot();
  });
});

describe('Init method', () => {
  test('shout call reset method', () => {
    const resetSpy = jest.spyOn(calc, 'reset');
    calc.init();
    expect(resetSpy).toHaveBeenCalled();
  });

  test('should call listenToButtons method', () => {
    const resetSpy = jest.spyOn(calc, 'listenToButtons');
    calc.init();
    expect(resetSpy).toHaveBeenCalled();
  });
});

describe('Call function after button is clicked', () => {
  let buttonSpyFn;

  beforeEach(() => {
    calc.init();
  });

  test('should call showInputNumbers after number button is clicked', () => {
    buttonSpyFn = jest.spyOn(calc, 'showInputNumbers');
    const numberBtn = document.querySelector('[data-button="number"]');

    simulateClick(numberBtn);

    expect(buttonSpyFn).toHaveBeenCalled();
    expect(buttonSpyFn).toHaveBeenCalledWith(numberBtn.textContent);
  });

  test('should call chooseOperator after operation button is clicked', () => {
    buttonSpyFn = jest.spyOn(calc, 'chooseOperator');
    const operatorBtn = document.querySelector('[data-button="operator"]');
    
    simulateClick(operatorBtn);
    
    expect(buttonSpyFn).toHaveBeenCalled();
    expect(buttonSpyFn).toHaveBeenCalledWith(operatorBtn.textContent);
  });

  test('should call equals function after equals button is clicked', () => {
    buttonSpyFn = jest.spyOn(calc, 'equals');
    const equalsBtn = document.querySelector('[data-button="equals"]');

    simulateClick(equalsBtn);

    expect(buttonSpyFn).toHaveBeenCalled();
  });

  test('should call reset function after clear button is clicked', () => {
    buttonSpyFn = jest.spyOn(calc, 'reset');
    const clearBtn = document.querySelector('[data-button="clear"]');
    
    simulateClick(clearBtn);
    
    expect(buttonSpyFn).toHaveBeenCalled();
  });

  // test('displays number after click', async () => {
  //   const btnOne = document.querySelector('#one');
  //   const currentInput = document.querySelector('#currentInput');
  //   await btnOne.click();
  //   expect(currentInput.textContent).toEqual('1');
  // });

  // test('should call equals function', () => {
  //   previousInput = 5;
  //   currentInput = 7;
  //   const equals = document.querySelector('#equals');
  //   equals.click();
  //   expect(equalsFn).toHaveBeenCalled();
  // });
});


describe('Calculate mathematics operations', () => {
  document.body.innerHTML = `<div id="previousInput">21</div>'
  + '<div id="currentInput">3</div>'
  + '<div id="sign">+</div>`;
  let previousInput;
  let currentInput;
  let signInput;
  let calc;

  test('schould be defined', () => {
    document.body.innerHTML = '<div id="previousInput">21</div>'
    + '<div id="currentInput">3</div>'
    + '<div id="sign">+</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);

    expect(calc.doMath).toBeDefined();
    expect(calc.doMath).not.toBeUndefined();
  });

  test('add 21 to 3', () => {
    document.body.innerHTML = '<div id="previousInput">21</div>'
    + '<div id="currentInput">3</div>'
    + '<div id="sign">+</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);

    expect(calc.doMath()).toBe(24);
  });

  test('add 17.2 to 3.5', () => {
    document.body.innerHTML = '<div id="previousInput">17.2</div>'
    + '<div id="currentInput">3.5</div>'
    + '<div id="sign">+</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);

    expect(calc.doMath()).toBe(20.7);
  });

  test('add -14 to 9', () => {
    document.body.innerHTML = '<div id="previousInput">-14</div>'
    + '<div id="currentInput">9</div>'
    + '<div id="sign">+</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);

    expect(calc.doMath()).toBe(-5);
  });

  test('subtract 12 from 7', () => {
    document.body.innerHTML = '<div id="previousInput">12</div>'
    + '<div id="currentInput">7</div>'
    + '<div id="sign">-</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(5);
  });

  test('subtract 55.5 from 50.1', () => {
    document.body.innerHTML = '<div id="previousInput">55.5</div>'
    + '<div id="currentInput">50.1</div>'
    + '<div id="sign">-</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(5.4);
  });

  test('multiply 4 by 11', () => {
    document.body.innerHTML = '<div id="previousInput">4</div>'
    + '<div id="currentInput">11</div>'
    + '<div id="sign">×</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(44);
  });

  test('multiply 16.2 by 0.5', () => {
    document.body.innerHTML = '<div id="previousInput">16.2</div>'
    + '<div id="currentInput">0.5</div>'
    + '<div id="sign">×</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(8.1);
  });

  test('divide 60 by 10', () => {
    document.body.innerHTML = '<div id="previousInput">60</div>'
    + '<div id="currentInput">10</div>'
    + '<div id="sign">÷</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(6);
  });

  test('divide 14 by 3', () => {
    document.body.innerHTML = '<div id="previousInput">14</div>'
    + '<div id="currentInput">3</div>'
    + '<div id="sign">÷</div>';
    previousInput = document.querySelector('#previousInput');
    currentInput = document.querySelector('#currentInput');
    signInput = document.querySelector('#sign');
    calc = new Calculator(previousInput, currentInput, signInput);
    expect(calc.doMath()).toBe(4.66666667);
  });
});

// test('Default value of calculator schould be zero', () => {
//   const currentValue = document.querySelector('#currentInput');
//   expect(currentValue.text()).toEqual('0');
// });

// const { assert } = require('console');
// const fs = require('fs')
// const $ = require('jquery');

// 'use strict';

// test('verify if test is visible and the color is correct', () => {
//     // given (settery, ustawimy test)
//     const data = fs.readFileSync(__dirname + '/../index.html', 'utf8')
//     document.body.innerHTML = data;

//     // when
//     const e = $('#main');

//     // then
//     expect(e.text()).toEqual('Witajcie Kochani');
//     expect(e.css('color')).toEqual('brown');
// });
