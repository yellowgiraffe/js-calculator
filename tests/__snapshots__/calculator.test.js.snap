// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class Calculator constructor should get DOM elements for display 1`] = `
<div
  class="current-input"
  id="currentInput"
>
  0
</div>
`;

exports[`Class Calculator constructor should get DOM elements for display 2`] = `
<div
  class="previous-input"
  id="previousInput"
/>
`;

exports[`Class Calculator constructor should get DOM elements for display 3`] = `
<div
  class="operation"
  id="sign"
/>
`;

exports[`Class Calculator constructor should get DOM elements for display 4`] = `
<div
  class="buttons"
  id="buttons"
>
  
        
  <button
    class="button sign-btn"
    data-button="operator"
    id="plus"
  >
    +
  </button>
  
        
  <button
    class="button sign-btn"
    data-button="operator"
    id="minus"
  >
    -
  </button>
  
        
  <button
    class="button sign-btn"
    data-button="operator"
    id="multiply"
  >
    ร
  </button>
  
        
  <button
    class="button sign-btn"
    data-button="operator"
    id="divide"
  >
    รท
  </button>
  

        
  <button
    class="button equals-btn"
    data-button="equals"
    id="equals"
  >
    =
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="seven"
  >
    7
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="eight"
  >
    8
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="nine"
  >
    9
  </button>
  

        
  <button
    class="button num-btn"
    data-button="number"
    id="four"
  >
    4
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="five"
  >
    5
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="six"
  >
    6
  </button>
  

        
  <button
    class="button num-btn"
    data-button="number"
    id="one"
  >
    1
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="two"
  >
    2
  </button>
  
        
  <button
    class="button num-btn"
    data-button="number"
    id="three"
  >
    3
  </button>
  

        
  <button
    class="button num-btn zero-btn"
    data-button="number"
    id="zero"
  >
    0
  </button>
  
        
  <button
    class="button"
    data-button="number"
    id="dot"
  >
    .
  </button>
  
        
  <button
    class="button"
    data-button="clear"
    id="clear"
  >
    C
  </button>
  
      
</div>
`;
